### PR TITLE
Karatsuba

### PR DESCRIPTION
**Pull request by definelicht**
Implement Karatsuba integer multiplication and use it for float multiplication.

### CODE REVIEW

File: device/Karatsuba.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
+#include "Karatsuba.h"
+
+#include <type_traits>  // std::enable_if
+
+template <int bits>
+typename std::enable_if<(bits > kMultBaseBits), ap_uint<2 * bits>>::type _Karatsuba(ap_uint<bits> const &a,
```
Comment (ChrisPattison) [inline device/Karatsuba.cpp]: I'm generally in favor of trailing return type for long function templates. It makes it a bit easier to read

File: device/Karatsuba.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                                                                                    ap_uint<bits> const &b) {
+#pragma HLS INLINE
+    using Full = ap_uint<bits>;
+    using Half = ap_uint<bits / 2>;
```
Comment (ChrisPattison) [inline device/Karatsuba.cpp:11]: Should we just guard that bits is even so we don't have to worry about bits/2 truncating?

File: device/MatrixMultiplication.cpp [hunk 3]
```diff
@@ -84,24 +92,78 @@ void Compute(hlslib::Stream<PackedFloat> &a_in, hlslib::Stream<PackedFloat> &b_i
...
+                const ap_uint<kBits> a_mantissa_padded = *reinterpret_cast<ap_uint<kMantissaBits> const *>(a.mantissa);
+                const ap_uint<kBits> b_mantissa_padded = *reinterpret_cast<ap_uint<kMantissaBits> const *>(b.mantissa);
+                // Meat of the computation. Only keep the top bits of the computation and throw away the rest
+                const ap_uint<kMantissaBits + 1> _m_mantissa =
+                    Karatsuba(a_mantissa_padded, b_mantissa_padded) >> (kMantissaBits - 1);
+                // We need to shift the mantissa forward if the most significant bit is not set
+                const bool should_be_shifted = !IsLastBitSet(_m_mantissa);
+                ap_uint<kMantissaBits> m_mantissa = should_be_shifted ? _m_mantissa : (_m_mantissa >> 1);
+                // The sign is just the XOR of the existing signs
+                const bool m_sign = (a.sign != 0) ^ (b.sign != 0);
```
Comment (ChrisPattison) [inline device/MatrixMultiplication.cpp]: There's a neat trick I use for integer xor: `a.sign != b.sign`

**Review by ChrisPattison — APPROVED** [review]
I left one or two comments that I don't feel particularly strongly about, so I leave it to you what you want to do with them

File: device/Karatsuba.cpp (hunk 2 above)
Comment (definelicht) [reply device/Karatsuba.cpp:11]: Done in 3c7f039784bb3025b4aa229e1542fabd7dda2e87

File: device/Karatsuba.cpp (hunk 1 above)
Comment (definelicht) [reply device/Karatsuba.cpp]: Done in 71cd775

File: device/MatrixMultiplication.cpp (hunk 3 above)
Comment (definelicht) [reply device/MatrixMultiplication.cpp]: Done in 9487682

**Pull request merged**